src/ErrorFormatter/AbstractFormatter.php

```diff
@@ -0,0 +1,60 @@
+<?php
+declare(strict_types = 1);
+
+namespace Middlewares\ErrorFormatter;
+
+use Middlewares\Utils\Factory;
+use Middlewares\Utils\HttpErrorException;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Throwable;
+
+abstract class AbstractFormatter implements FormatterInterface
+{
+    protected $responseFactory;
+    protected $contentTypes = [];
+
+    public function __construct(
+        ResponseFactoryInterface $responseFactory = null
+    ) {
+        $this->responseFactory = $responseFactory ?? Factory::getResponseFactory();
+    }
+
+    public function isValid(Throwable $error, ServerRequestInterface $request): bool
+    {
+        $accept = $request->getHeaderLine('Accept');
+
+        foreach ($this->contentTypes as $type) {
+            if (stripos($accept, $type) !== false) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    abstract protected function format(Throwable $error): string;
+
+    public function handle(Throwable $error, ServerRequestInterface $request): ResponseInterface
+    {
+        $response = $this->responseFactory->createResponse($this->errorStatus($error));
+        $response = $response->withHeader('Content-Type', $this->contentTypes[0]);
+        $response->getBody()->write($this->format($error));
```

Comment:
There is no possible way to enforce that the stream is empty and/or writable. This should be using `StreamFactoryInterface`, IMO.

Reply:
Does not PSR-17 ensure a response recently created is writable?
Does this means every time I create a response, I must discard the existing body and create a new one?

Reply:
There is no way to enforce it by code. It can only document the requirement. Replacing the body is the safest way to ensure that the response (or request) is in a consistent state.

Reply:
There's no way to enforce the stream factory by code too.
If a body is not writable, it should throw a runtime exception, saying the factory is not valid. Trying to solve it in a hidden way, requiring a new dependency is a bad idea, IMO.

Reply:
I got the same vision as oscarotero. All of the PSR7/17 implementations I know create a response with a writable steam. And per the spec, the write() function should throw a runtimeexcetion when it couldn't write.
I think it's pretty safe to use directly the write function in your code without replacing the body. But it's only my opinion.

Reply:
I think this issue does not affect here, because the new stream is created empy (no from a string or file).

Reply:
That the stream is created empty is only an assumption, there is no way to prove it without checking the size of the stream, which may or may not be supported.